ERCS/erc-7678.md

```diff
@@ -0,0 +1,223 @@
+}
+```
+
+### Using the UserOp builder
+
+To build a UserOp using the UserOp builder, the building party should proceed as follows:
+
+1. Obtain the address of `UserOpBuilder` and a `context` from the account owner.  The `context` is an opaque bytes array from the perspective of the building party.  The smart account implementation may need the `context` in order to properly figure out the UserOp fields.  See the “Rationale” section for more info.
+2. Execute a multicall (batched `eth_call`s) of `getNonce`, `getCallData`, `getDummySignature` with the `context` and executions.  The building party will now have obtained the nonce, calldata, and dummy signature (see “Rationale” for what a dummy signature is).
+3. Fill out a UserOp with the data obtained previously.  This UserOp must be valid except for the `signature` field.  Then call (via `eth_call`) `getSignature` with the UserOp and `context` to obtain a completely valid UserOp.
+    1. Note that a UserOp has a lot more fields than `nonce`, `callData`, and `signature`, but how the building party obtains the other fields is outside of the scope of this document, since only these three fields are heavily dependent on the smart account implementation.
+
+At this point, the building party has a completely valid UserOp that they can then submit to a bundler or do whatever it likes with it.
+
+### Using the UserOp builder when the account hasn’t been deployed
+
+If the account has yet to be deployed, which means that the building party is looking to send the very first UserOp for this account, then the building party may modify the flow above as follows:
+
+- In addition to the `UserOpBuilder` address and the `context`, the building party also obtains the `factory` and `factoryData` as defined in ERC-4337.
+- When calling one of the view functions on the UserOp builder, the building party may use `eth_call` to deploy the `CounterfactualCall` contract with `factory` and `factoryData` (see below).  The `CounterfactualCall` contract would deploy the account before calling the view functions on the UserOp builder.
+- When filling out the UserOp, the building party includes `factory` and `factoryData`.
+
+### Counterfactual call
+
+The counterfactual call contract is inspired by [ERC-6492](./eip-6492.md), which devised a mechanism to execute `isValidSignature` (see [ERC-1271](./eip-1271.md)) against a pre-deployed (counterfactual) contract.
+
+```solidity
+contract CounterfactualCall {
+    
+    error CounterfactualDeployFailed(bytes error);
+
+    constructor(
+        address smartAccount,
+        address create2Factory, 
+        bytes memory factoryData,
+        address userOpBuilder, 
+        bytes memory userOpBuilderCalldata
+    ) { 
+        if (address(smartAccount).code.length == 0) {
+            (bool success, bytes memory err) = create2Factory.call(factoryData);
+            if (!success) revert CounterfactualDeployFailed(err);
+        }
+
```

Comment:
it's really important to check again for  `if (address(smartAccount).code.length == 0)` here because the call may have succeeded but not deployed the correct contract (or not deployed anything at all), which will cause a silent fail later on, as a call to an EOA will just succeed.

Reply:
CC @filmakarov

Reply:
suggested a change below.
Thank you @Ivshti

Reply:
@arein I think this can be marked as resolved?